### PR TITLE
Support partial merge pushdown 

### PR DESCRIPTION
**Pull request by wenleix**
## Introduction 

Grouped execution was introduced to Presto in https://github.com/prestodb/presto/pull/8951 to support huge join and aggregation raised in ETL pipelines. 

A large scale warehouse often optimize table layout by pre-bucketing tables on frequently-used join/aggregation keys to avoid the cost of data re-partitioning. However, tables often have different (yet compatible) number of buckets. 

For example, an extremely huge table A might have 8192 buckets, while another large table have 1024 buckets. A full repartitioning for queries like A JOIN B is undesired because:

- In-memory repartitioning loses the benefit of grouped execution and the query may exceeded maximum distributed memory allowed.
- Materialized exchange (https://github.com/prestodb/presto/issues/12387) will help run the query within memory limit, however it will be very inefficient in terms of CPU and IO. 

https://github.com/prestodb/presto/pull/11749 is the first work  try to optimize this problem, and lay some fundamental building blocks to this problem. However, in practice we found the following issue impedes it from a general useful solution: (e.g. we have to ask users to opt-in, instead of enable it by default). 

- It couples partitioning selection and partial merge pushdown. So it can change how partitioning is chosen and  cause query to fail. For example:

```
          AGGR
            |
          JOIN
         /     \
 TS(2048)    TS(1024)
```

`AddExchange` always prefer to use left side partitioning if exists. In this case, it will cause less number of lifespan for aggregation and increase the global memory usage. 

- The common partitioning selection logic is opaque to engine. Thus engine cannot reason about whether this partial merge pushdown is safe or not, for example:

```
          JOIN
         /     \
 TS(1024)    AGGR(2048)
                    \
                   TS(2048)
```

Push partial merge to the right TS can cause OOM. However, since engine doesn't know whether the common partitioning returned by Connector is a refinement or coarsening, it cannot reason which partitioning to use. 

## Partial Repartitioning  and Partial Merge

This PR explores an alternative attempt based on the supports built in https://github.com/prestodb/presto/pull/11749 . The idea is to efficiently support these partitioning adaptions, we need the two "partial repartitioning" and "partial merge"  exchange types introduced in SCOPE paper ([1]):  

Screen Shot 2019-04-07 at 8 45 10 PM

Consider the same example above `A (8192) JOIN B (1024)`.  Instead of do a full re-partitioning over A or B, we can do a "partial merge" over A into 1024 buckets (we can also do "partial re-partitioning" over B, but this is beyond the scope of this PR).   

## Implementation

We are going to discuss about implementation of Partial Repartitioning  and Partial Merge in the context of materialized exchange. In memory exchange would be more interesting to discuss. 

### Partial Repartitioning

Partial repartitioning implementations is more straightforward. Since it's always going to be like this in the materialized exchange context:

```
      TableWriter
            |
      Exchange (partial repartitinoing)
            |
      ...............
``` 

Instead of a remote exchange, the `ExchangeNode` can be simply expanded to a local exchange operator.  It can even be pushed into `TableWriter` if parallelism is not a concern.

### Partial Merge

Partial merge is more interesting since it can be handled in different ways.  For the convenience of discussion, we assuming we are trying to do partial merge from 4096 buckets to 1024 buckets. 

- Directly supporting it in Presto engine requires a bit more work, potential solutions:
    * We can use a local exchange operator to "merge" the data produced from different lifespans, but the engine needs to intelligently schedule lifespans to be coalesced in the same time and on the same nodes. 
    * We can allow partial writes to a bucket (e.g. doing a partial merge from 4096 buckets to 1024 buckets, one lifespan can finish but the target output bucket is only 25% finished)

```
      TableWriter
            |
      Exchange (partial merge)
            |
      ...............
``` 

- It can be fully eliminated by pushing down partial merge to the table scan. Note pushing through join/aggregation has the risk to increase memory usage.

- If it cannot be pushed down, another way would be doing "partial merge" in a lazy fashion. i.e. we still writes to 4096 buckets, but in the next table scan, it reads out as 1024 buckets. 

Note partial merge can be conveniently pushdown to Connector by asking an "alternative TableLayout". 

[1]: http://www.cs.columbia.edu/~jrzhou/pub/Scope-VLDBJ.pdf

## Misc

The idea here is if exchange materialization is considered as the primary way to support large queries; the problem is much simplified, since all partitioning will remain the same during query execution (same number of buckets as specified by `num_hash_partitions`). 

In this world, two major primitives (besides full repartitioning) are "partial merge" and "partial repartitioning" which used at beginning to adapt query specified partition count. And they can be modeled as "partial merge pushdown" and "partial repartitioning pushup".  

Compatible partitioning is supported as a by-product. 

## Open Questions

How do Hive/Spark handles "compatible buckets"?

### CODE REVIEW

**Comment by wenleix** [comment]
cc @arhimondr , @rschlussel, @hellium01

File: presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java [hunk 1]
```diff
@@ -92,6 +92,35 @@
      */
     Optional<PartitioningHandle> getCommonPartitioning(Session session, PartitioningHandle left, PartitioningHandle right);
 
+    /**
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java]: I wonder if it makes sense to copy this long comment. What if we leave it only in the SPI in the  `ConnectorMetadata` class?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 2]
```diff
@@ -2470,6 +2473,106 @@ public void testMismatchedBucketing()
         }
     }
 
+    @Test
+    public void testPartialMergePushdown()
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Could you please run it with and without materialization enabled? (see buckete execution tests)

Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Could you please run it with and without materialization enabled? (see buckete execution tests)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 3]
```diff
@@ -550,7 +555,12 @@ else if (redistributeWrites) {
                 }
             }
 
-            if (partitioningScheme.isPresent() && !source.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, metadata, session)) {
+            if (partitioningScheme.isPresent() &&
+                    // TODO: Deprecate compatible table partitioning
+                    !source.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, metadata, session) &&
+                    !(partialMergePushdownStrategy == PUSHDOWN_BUT_NOT_THROUGH_JOIN_AND_AGGREGATION &&
+                            source.getProperties().isRefinedPartitioningOver(partitioningScheme.get().getPartitioning(), false, metadata, session) &&
+                            canPushdownPartialMerge(source.getNode()))) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Just to make sure I understand it correctly

We only optimizing simple projections/filters of a bucketed table into the other bucketed table? It doesn't save any memory, but potentially saves CPU cycles on doing shuffle.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java [hunk 4]
```diff
@@ -123,6 +123,44 @@ public boolean isCompatibleWith(
         return true;
     }
 
+    public boolean isRefinedPartitioningOver(
+            Partitioning right,
+            Metadata metadata,
+            Session session)
+    {
+        if (!handle.equals(right.handle) && !metadata.isRefinedPartitioningOver(session, handle, right.handle)) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: `handle.equals(right.handle)` - do you think we should leave a comment (or give a better name to the method) to reflect that by `Refined` we mean `SameOrRefined`?  I'm not sure if it is worth the verbosity though.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 5]
```diff
@@ -72,6 +72,7 @@
     private boolean spatialJoinsEnabled = true;
     private boolean fastInequalityJoins = true;
     private JoinReorderingStrategy joinReorderingStrategy = ELIMINATE_CROSS_JOINS;
+    private PartialMergePushdownStrategy partialMergePushdownStrategy = PartialMergePushdownStrategy.NONE;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: It should be save to have it on by default.

File: presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java (hunk 1 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java]: @arhimondr : I agree we should only keep in one place and have a de-briefed version in other place. Just don't know where should be the "full" version be. 

Any suggestions @rschlussel  ? :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java (hunk 4 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: I originally start with `isRefinedPartitioningOverOrEqual`,  but then personally feel it's too wordy...

I think we can have comments state that "RefinedOver` is reflective.  
 
Thoughts? @rschlussel

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java (hunk 3 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: @arhimondr  Correct. For now it's mainly for completeness with some limited usage (read and directly write to a bucketed table with less number of buckets.

In the future, thinking about we can have more aggressive `ALWAY_PUSHDOWN_PARTIAL_MERGE` as `PartialMergePushdownStrategy` -- it might be useful in some use case (at the cost of more CPU...)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java (hunk 5 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: @arhimondr : Still we want to gradually roll out right? :)

**Comment by wenleix** [comment]
@arhimondr 

> > Mark common partitioning API as deprecated 

> Why don't we simply get rid of it?

It's used by `hive.optimize-mismatched-bucket-count`. So we will have to deprecate it first..

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 6]
```diff
@@ -153,6 +154,12 @@ public boolean canReplicate()
         }
     }
 
+    public enum PartialMergePushdownStrategy
+    {
+        NONE,
+        PUSHDOWN_BUT_NOT_THROUGH_JOIN_AND_AGGREGATION
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: Any thoughts about the name (looks long but I don't know how to make it better...) ? @rschlussel

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 7]
```diff
@@ -551,7 +556,12 @@ else if (redistributeWrites) {
                 }
             }
 
-            if (partitioningScheme.isPresent() && !source.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, metadata, session)) {
+            if (partitioningScheme.isPresent() &&
+                    // TODO: Deprecate compatible table partitioning
+                    !source.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, metadata, session) &&
+                    !(partialMergePushdownStrategy == PUSHDOWN_BUT_NOT_THROUGH_JOIN_AND_AGGREGATION &&
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: could this whole thing be captured in canPushDownPartialMerge? Right now here and a few times below you have to separately capture all the logic about when partialmerges get pushed down, which it more confusing and error prone.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 8]
```diff
@@ -1839,6 +1839,25 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
...
+            return false;
+        }
+
+        int leftBucketCount = leftHandle.getBucketCount();
+        int rightBucketCount = rightHandle.getBucketCount();
+        return leftBucketCount == rightBucketCount || // must be evenly divisible
+                (leftBucketCount % rightBucketCount == 0 &&
+                        // ratio must be power of two
+                        // TODO: this can be relaxed
+                        Integer.bitCount(leftBucketCount / rightBucketCount) == 1);
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @rschlussel : No fundamental reason. But when @haozhun  initially implementing bucket adaption for Hive connector, he chose only to supports for power of two. So for now let's enforce this :)
